app/services/arcgis_api/response_validation.rb

```diff
@@ -0,0 +1,28 @@
+module ArcgisApi
+  # Faraday middleware to raise exception when response status is 200 but with error in body
+  class ResponseValidation < Faraday::Middleware
+    def on_complete(env)
+      return unless env[:status] == 200 && env.body
+      body = env.body.is_a?(String) ? JSON.parse(env.body) : env.body
+      return unless body.fetch('error', false)
+      handle_api_errors(body)
+    rescue => e
+      raise e if e.is_a?(Faraday::ServerError)
+    end
+
+    def handle_api_errors(response_body)
+      # response_body is in this format:
+      # {"error"=>{"code"=>400, "message"=>"", "details"=>[""]}}
+      error_code = response_body.dig('error', 'code')
+      error_message = response_body.dig('error', 'message') || "Received error code #{error_code}"
+      # log an error
+      raise Faraday::ServerError.new(
+        RuntimeError.new(error_message),
+        {
+          status: error_code,
+          body: { details: response_body.dig('error', 'details')&.join(', ') },
+        },
+      )
```

Comment:
Is this only going to be reached on 5xx errors?

Reply:
Since this isn't a 5xx, I don't think we should use `Faraday::ServerError`. We can use a custom error type instead and then if needed [update the exceptions configured with `faraday-retry`](https://github.com/lostisland/faraday-retry#specify-which-exceptions-should-trigger-a-retry).